[Feature] Theme manager

roya0045 opened the pull request:
## Description
As proposed in https://github.com/qgis/QGIS-Enhancement-Proposals/issues/211

As discussed the goal of this is to provide a better interface to interact with existing themes and their content to sort out errors and easily correct them without going the select->modify->replace route

The widget has theme browsing arrows and combo-box at the top, a dedicated layertreeview in the center and buttons to import, remove layers and to create and delete themes.

The import and remove buttons are voluntarily bigger as they should be more commonly used.

![theme_manager](https://user-images.githubusercontent.com/12854129/113455453-a5dee780-93d8-11eb-97ee-16879f6662a6.PNG)

GIF showcase (forgot to show the context menu to unhide non-layer items)
![theme_manager](https://user-images.githubusercontent.com/12854129/115561081-38113780-a283-11eb-9482-4a36dd5f7c19.gif)
### Here is a list of the changes and their reasons:
#### QgsMapThemes: 

-  Exposed the addPreset slot to use the existing dialog when creating a new theme

#### QgsLayerTreeViewer:

- Added fid based filtering to the QgsLayerTreeProxyModel, needed to only display layers within a theme

#### QgsThemeViewer:

-  Customized QgsLayerTreeView with specific dedicated drag&drop behaviour, had to be made separate to prevent circular dependencies.
-  Added custom slot handling to add selected layers from the layer view to the current theme
- Added custom code to prevent moving/remove layer from the project via bad transfer, prevent drag and drop and ease selection of layers.
- Also allow the control of layers (display or not, etc) from the widget to inspect elements. But does not allow reordering ( to prevent drag and drop issues).

#### QgsThemeManagerWidget:

- Base class with the logic and widgets
- Browse existing themes
- Add selected layers to a theme via drag & drop or with the 'import' button
- Remove selected layer from a theme
- Create new theme from current view (see changes in QgsMapTheme)
- Delete current theme

#### QgsLayerTreeView
 - Added a variable to the setModel method, otherwise multiple view connected to the same model would crash the model when a row was added as the secondary view would fetch invalid data to populate another row.

roya0045 commented:
I'll try to provide some images and gif (if needed) later.

roya0045 commented:
@elpaso do you know why ming64 is failing, I also tried to run it once on a branch with all my PR in my fork and couldn't get it to work.

roya0045 commented:
Seem to get random crash, unsure what may cause this, need to investigate.

roya0045 commented:
Ok things left to solve:

- Adding or removing themes trigger the combobox update, but only the previous items are fetched. Possible issues with the mapthemecollection?
- Crash when changing symbology. Occurs in QgsLayerTreeView. @nyalldawson I'd be open to hear if you have a suggestion. 
- [Extra] Groups, even when empty seem to be considered as layers in the record and have an ID, thus not filtered by the proxy, need to asses if fixing is problematic or out of scope.

roya0045 commented:
Here is the stack trace caused by adding a symbol, crash is semi-consistent.

> 1  0x00005584ba21c754 in qgisCrash(int) ()
2  0x00007f8e5ce9e950 in <signal handler called> () at /lib/x86_64-linux-gnu/libc.so.6
3  0x00007f8e628974dc in QgsLayerTreeNode::parent() () at /home/ar/Documents/qgis/build/output/lib/libqgis_app.so.3.19.0
4  0x00007f8e5dbf8165 in QgsLayerTreeModel::node2index(QgsLayerTreeNode*) const () at /home/ar/Documents/qgis/build/output/lib/libqgis_core.so.3.19.0
5  0x00007f8e61644055 in QgsLayerTreeView::node2sourceIndex(QgsLayerTreeNode*) const () at /home/ar/Documents/qgis/build/output/lib/libqgis_gui.so.3.19.0
#6  0x00007f8e61643fda in QgsLayerTreeView::node2index(QgsLayerTreeNode*) const () at /home/ar/Documents/qgis/build/output/lib/libqgis_gui.so.3.19.0
7  0x00007f8e61642691 in QgsLayerTreeView::updateExpandedStateFromNode(QgsLayerTreeNode*) () at /home/ar/Documents/qgis/build/output/lib/libqgis_gui.so.3.19.0
8  0x00007f8e61641914 in QgsLayerTreeView::modelRowsInserted(QModelIndex const&, int, int) () at /home/ar/Documents/qgis/build/output/lib/libqgis_gui.so.3.19.0
9  0x00007f8e61648c48 in QtPrivate::FunctorCall<QtPrivate::IndexesList<0, 1, 2>, QtPrivate::List<QModelIndex const&, int, int>, void, void (QgsLayerTreeView::*)(QModelIndex const&, int, int)>::call(void (QgsLayerTreeView::*)(QModelIndex const&, int, int), QgsLayerTreeView*, void**) () at /home/ar/Documents/qgis/build/output/lib/libqgis_gui.so.3.19.0
10 0x00007f8e616481e7 in void QtPrivate::FunctionPointer<void (QgsLayerTreeView::*)(QModelIndex const&, int, int)>::call<QtPrivate::List<QModelIndex const&, int, int>, void>(void (QgsLayerTreeView::*)(QModelIndex const&, int, int), QgsLayerTreeView*, void**) () at /home/ar/Documents/qgis/build/output/lib/libqgis_gui.so.3.19.0
11 0x00007f8e6164729b in QtPrivate::QSlotObject<void (QgsLayerTreeView::*)(QModelIndex const&, int, int), QtPrivate::List<QModelIndex const&, int, int>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) () at /home/ar/Documents/qgis/build/output/lib/libqgis_gui.so.3.19.0
12 0x00007f8e5fa942e6 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
13 0x00007f8e5f9f7f82 in QAbstractItemModel::rowsInserted(QModelIndex const&, int, int, QAbstractItemModel::QPrivateSignal) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
14 0x00007f8e5fa004f2 in QAbstractItemModel::endInsertRows() () at /lib/x86_64-linux-gnu/libQt5Core.so.5

roya0045 commented:
Solved the parent issue by adding some safety to the function to prevent passing a corrupted pointer. Did the trick, now get crashes because the expanded variable is not initiated when querying. I have no clues why the normal layertreeview doesn't have these issues and disabling filtering on the proxy didn't have any effect on the crashes.

roya0045 commented on src/gui/layertree/qgslayertreeview.cpp:
```diff
@@ -95,10 +95,11 @@ void QgsLayerTreeView::setModel( QAbstractItemModel *model )
-
+  if ( connectProxy )
+  {
+    connect( mProxyModel, &QAbstractItemModel::rowsInserted, this, &QgsLayerTreeView::modelRowsInserted );
```
@nyalldawson This was the cause of the crashes, upon adding a legend node in the main view this would also be called in the custom view, in the instances where the model didn't have any 'spare' rows ( no nodes removed previously or other state), this would fetch data from an invalid node (as the good one would be added already). This is why it lead to crashes. I made the parameter optional to maintain everything functionnal.

roya0045 commented:
@andreasneumann @mach0  I added a gif, I thought this might interest you.

roya0045 commented:
Group filtering is not perfect yet. If groups are hiddent, items within are also hidden. I'll try to find a way to deal with this.

olivierdalang commented:
I'm a bit late in the game (sorry missed the QEP), but was wondering whether we've considered other UI approaches that don't involve yet another dock widget. The proposed approach takes a lot of screen space, mostly for repeating information that's already on screen. This means users would always end up toggling that panel on and off.

Wouldn't it be possible to just add the theme switcher dropdown/arrows at the top of the legend ? By just adding apply/update buttons, you get 90% of the functionality, but much better UX in my opinion.

For reference, long time ago (deprecated since themes where introduced into QGIS) I had made a LayerCombination plugins working on that principle : https://vimeo.com/70446444

Partial themes (themes that define visibility/invisibility only for some layers, leaving the other as is) would be a bit harder to implement with such an UI, but do we really need that feature ? If yes, may be we could achieve this in the main legend using the underused QgsLayerTreeViewIndicator instead.

roya0045 commented:
@olivierdalang The intent of a duplicate widget is to be able to have a clear picture of the content of a theme and to be able to easily add layers from the project. Viewing is one aspect but being able to easily add and remove item is also important. Doing this in the legend currently is more complex and more prone to user error. 

Doing any sort of manipulation with a single layertreeview becomes unwieldy and finnicky. The current proposal allows the user to be able to easily change the content of any themes without losing any information.

We have a panel for Tile which is a single slider so I don't see why this must be minimized considering the current precedent.

And yes dealing with partial themes is not easy since themes are manager at layer level. The other current issue is groups. Properly hiding them would required checking all nested layers and subgroup to ensure none are needed and then checking them again once they are evaluated individually. But this would not change either way.

Though if you have suggestions or ideas other than trying to push this on the main layertreeview, that would be appreciated. 

I initially considered a plugin but Mr. Newman indicated that such a feature could be useful in the project.

roya0045 commented:
Not stale.

roya0045 commented:
fine

roya0045 commented:
staying alive

roya0045 commented:
its fine

roya0045 commented:
awaiting review/exchange

mach0 commented:
while I am not the one to review (but one who might add some features later to this request) I am also waiting for inclusion into core ..

roya0045 commented:
> while I am not the one to review (but one who might add some features later to this request) I am also waiting for inclusion into core ..

Thank you for your support, out of curiosity have you tested the feature ? I'm curious to get some feedback as I tend to merge to all my pending PR in a build and use it a work, I don't get much suggestions from my colleagues.

mach0 commented:
@roya0045 yes I did compile and test your branch and everything I tested worked as expected.
But I am going to checkout and build again today evening to maybe implement a few more functions I would need too.

roya0045 commented:
> @roya0045 yes I did compile and test your branch and everything I tested worked as expected.
> But I am going to checkout and build again today evening to maybe implement a few more functions I would need too.

You are welcome to commit to this PR!

mach0 commented:
hmm - not much going on here .. still in the slow lane ..

roya0045 commented:
staying alive

mach0 commented:
holding this up

roya0045 commented:
Something seems wonky with groups, I'll try to fix it soon.

roya0045 commented:
Groups containing layers and their visible layers should now be displayed properly. The context action was changed to display empty group and non-spatial layer if needed.

roya0045 commented:
Test failures seem unrelated.

roya0045 commented:
Not stale

roya0045 commented:
PR is still valid.

roya0045 commented:
it's fine

roya0045 committed:
Layer level theme-manager

Could be done on symbol level, would require more important rework of the layertree.

roya0045 committed:
add note to self

baswein commented:
I use themes all the time and am really excited about this!

roya0045 commented:
> I use themes all the time and am really excited about this!

You can try it if you compile a version with it, I tried to make a more in-depth version but I'm hitting roadblocks. I'm unsure if and when I'll reopen my closed branch as I'm using my work in my own custom build.

roya0045 commented:
@baswein I made a new build for testing some things, if you wanted to use the content of this branch on window you can get the artefact at https://github.com/roya0045/QGIS_1/actions/runs/2272082461 it should be up for 4 or 3 more days.

roya0045 commented:
Though it has issues loading projects, and I have no clue why, I assume it comes from a change in master but I have no idea what.

baswein commented:
Thanks - I'll give it a try.

roya0045 commented:
> Thanks - I'll give it a try.

To load projects, drag and drop them in, I have no clue why the normal loading fails, I'll dig deeper, it's also inconsistent. I'll try to see if I can fix it soon.

baswein commented:
Alex- This was really fun to test out. 

It worked very well for regular layers and I haven't had any problem loading projects.  

However if I have "Theme 1" up in the Theme manager and I turn a layer off in the layers panel it also toggles it in theme I have displayed in the theme manager (Theme 1). I feel like the Theme should only change when I turn layers on and off in the Theme manager not in the Layers Panel. 

Buttons I would like: 
1. Rename Theme
2. Apply Theme (set the theme as active in the layers panel)

I do think  the ability to choose the style (from the style manager) that is being applied would be important. Otherwise if you use styles you have to go back to the old replace theme method to manage the themes.

I can image it being either the style manager symbol next to the layer if the layer has named styles. 
![image](https://user-images.githubusercontent.com/15817745/166972630-454ce0d6-aca9-4004-9ceb-aee65980df52.png)
That you can click on and choose the named style.  

Or they show up below the layer similarly to the way categorized values show up below layers in the layer manager with mutual exclusive check-boxes to select the style to use.

roya0045 commented:
@baswein Thank you for the feedback! The aspect where the layers changed in the theme manager is intended. It is basically a filtered view of the main layertree. What you see are what is in the theme, you can add or remove the layers from the theme without breaking your current work-state. You can add layers by selecting them and hitting the import button or by drag & drop. 

The main limitation is for symbol and when using alternate sytle, I tried to work that around in the new alternate version but can't get it to work properly, it just crashes. The other option would be to have a duplicate model but that may break interactions, I would need some help or to dig deeper to find the best way to keep everything properly linked.

baswein commented:
Ok - that is not intuitive to me. I might come around to it but right now I feel like I want it to be:

- In the Layers Panel I control the visibility of the layers in the map. 
- And in the theme manager I control the visibility of the layers in the theme.

I think that would be a very clear separation to me. Otherwise I could very easily see cases where I have forgotten that I have the Theme manager open and I am toggling layers on and off in my layers panel as I do when I'm working on a map and then realize that I have inadvertently changed my carefully constructed visibility of whatever theme I happened to have open in the manager. I think this is why there is a warning about overwriting a theme because you don't want to do it inadvertently.     

I do like the filtered view and the drag and drop functionality. I could also see plus and minus buttons like there are in the layout legend panel. 

Wild Idea: 
How about a rename layer that allows you to have different names for layers that would be used in the layout legend depending on the theme.

roya0045 commented:
@baswein there are buttons, as for the rename, I'm not sure what you mean but that's touching too many things.

The main goal of the panel was to quickly browse through themes and to be able to identify missing or extra element and remove them, withouth having to apply the theme. That explain some of the design decisions.
